File: OutlineText/TextDesignerCSLibrary/TextGradOutlineStrategy.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Drawing;
using System.Drawing.Drawing2D;
using System.Drawing.Imaging;

namespace TextDesignerCSLibrary
{
    public class TextGradOutlineStrategy : TextImplGetHeight
	{
		public TextGradOutlineStrategy()
		{
			m_nThickness=2;
			m_brushText = null;
			m_bClrText = true;
            m_bUseCurvedGradient = false;
            disposed = false;
		}
        public override void Dispose()
		{
			Dispose(true);
			GC.SuppressFinalize(this);
		}
        protected override void Dispose(bool disposing)
		{
			if (!this.disposed)
			{
				if (disposing)
				{
                    if (m_brushText != null)
                    {
                        m_brushText.Dispose();
                    }
				}

				disposed = true;
			}
		}
		~TextGradOutlineStrategy()
		{
			Dispose(false);
		}
        public override ITextStrategy Clone()
		{
			TextGradOutlineStrategy p = new TextGradOutlineStrategy();
			if (m_bClrText)
                p.Init(m_clrText, m_clrOutline1, m_clrOutline2, m_nThickness, m_bUseCurvedGradient);
            else
                p.Init(m_brushText, m_clrOutline1, m_clrOutline2, m_nThickness, m_bUseCurvedGradient);

            return (ITextStrategy)(p);
		}

        public void Init(
            System.Drawing.Color clrText,
            System.Drawing.Color clrOutline1,
            System.Drawing.Color clrOutline2,
            int nThickness,
            bool useCurveGradient)
        {
            m_clrText = clrText;
            m_bClrText = true;
            m_clrOutline1 = clrOutline1;
            m_clrOutline2 = clrOutline2;
            m_nThickness = nThickness;
            m_bUseCurvedGradient = useCurveGradient;
        }

        public void Init(
            System.Drawing.Brush brushText,
            System.Drawing.Color clrOutline1,
            System.Drawing.Color clrOutline2,
            int nThickness,
            bool useCurveGradient)
        {
            m_brushText = brushText;
            m_bClrText = false;
            m_clrOutline1 = clrOutline1;
            m_clrOutline2 = clrOutline2;
            m_nThickness = nThickness;
            m_bUseCurvedGradient = useCurveGradient;
        }

        public override bool DrawString(
			System.Drawing.Graphics graphics,
			System.Drawing.FontFamily fontFamily,
			System.Drawing.FontStyle fontStyle,
			int fontSize,
			string strText,
			System.Drawing.Point ptDraw,
			System.Drawing.StringFormat strFormat)
		{
			using (GraphicsPath path = new GraphicsPath())
			{
				path.AddString(strText, fontFamily, (int)fontStyle, fontSize, ptDraw, strFormat);

				List<Color> list = new List<Color>();
                if (m_bUseCurvedGradient)
                    TextGradOutlineLastStrategy.CalculateCurvedGradient(m_clrOutline1, m_clrOutline2, m_nThickness, list);
                else
                    TextGradOutlineLastStrategy.CalculateGradient(m_clrOutline1, m_clrOutline2, m_nThickness, list);

                for (int i = m_nThickness; i >= 1; --i)
				{
					using (Pen pen1 = new Pen(list[i - 1], i))
					{
						pen1.LineJoin = LineJoin.Round;
						graphics.DrawPath(pen1, path);
					}
				}

				if (m_bClrText)
				{
					using (SolidBrush brush = new SolidBrush(m_clrText))
					{
						graphics.FillPath(brush, path);
					}
				}
				else
					graphics.FillPath(m_brushText, path);
			}
			return true;
		}


        public override bool DrawString(
			System.Drawing.Graphics graphics,
			System.Drawing.FontFamily fontFamily,
			System.Drawing.FontStyle fontStyle,
			int fontSize,
			string strText,
			System.Drawing.Rectangle rtDraw,
			System.Drawing.StringFormat strFormat)
		{
			using (GraphicsPath path = new GraphicsPath())
			{
				path.AddString(strText, fontFamily, (int)fontStyle, fontSize, rtDraw, strFormat);

				List<Color> list = new List<Color>();
                if (m_bUseCurvedGradient)
                    TextGradOutlineLastStrategy.CalculateCurvedGradient(m_clrOutline1, m_clrOutline2, m_nThickness, list);
                else
                    TextGradOutlineLastStrategy.CalculateGradient(m_clrOutline1, m_clrOutline2, m_nThickness, list);

                for (int i = m_nThickness; i >= 1; --i)
				{
					using (Pen pen1 = new Pen(list[i - 1], i))
					{
						pen1.LineJoin = LineJoin.Round;
						graphics.DrawPath(pen1, path);
					}
				}

				if (m_bClrText)
				{
					using (SolidBrush brush = new SolidBrush(m_clrText))
					{
						graphics.FillPath(brush, path);
					}
				}
				else
					graphics.FillPath(m_brushText, path);
			}
			return true;
		}

        public override bool GdiDrawString(
			System.Drawing.Graphics pGraphics,
			LOGFONT pLogFont,
			string pszText,
			System.Drawing.Point ptDraw)
		{
            using (GraphicsPath pPath = new GraphicsPath(System.Drawing.Drawing2D.FillMode.Winding))
            {
                bool b = GDIPath.GetStringPath(
                    pGraphics,
                    pPath,
                    pszText,
                    pLogFont,
                    ptDraw);

                if (false == b)
                {
                    return false;
                }

                List<Color> list = new List<Color>();
                if (m_bUseCurvedGradient)
                    TextGradOutlineLastStrategy.CalculateCurvedGradient(m_clrOutline1, m_clrOutline2, m_nThickness, list);
                else
                    TextGradOutlineLastStrategy.CalculateGradient(m_clrOutline1, m_clrOutline2, m_nThickness, list);

                for (int i = m_nThickness; i >= 1; --i)
                {
                    System.Drawing.Color clr = list[i - 1];
                    using (Pen pen = new Pen(clr, i))
                    {
                        pen.LineJoin = LineJoin.Round;
                        pGraphics.DrawPath(pen, pPath);
                    }
                }

                if (m_bClrText)
                {
                    using (SolidBrush brush = new SolidBrush(m_clrText))
                    {
                        pGraphics.FillPath(brush, pPath);
                    }
                }
                else
                {
                    pGraphics.FillPath(m_brushText, pPath);
                }
            }
			return true;
		}

        public override bool GdiDrawString(
			System.Drawing.Graphics pGraphics,
			LOGFONT pLogFont,
			string pszText,
			System.Drawing.Rectangle rtDraw)
		{
            using (GraphicsPath pPath = new GraphicsPath(System.Drawing.Drawing2D.FillMode.Winding))
            {
                bool b = GDIPath.GetStringPath(
                    pGraphics,
                    pPath,
                    pszText,
                    pLogFont,
                    rtDraw);

                if (false == b)
                {
                    return false;
                }

                List<Color> list = new List<Color>();
                if (m_bUseCurvedGradient)
                    TextGradOutlineLastStrategy.CalculateCurvedGradient(m_clrOutline1, m_clrOutline2, m_nThickness, list);
                else
                    TextGradOutlineLastStrategy.CalculateGradient(m_clrOutline1, m_clrOutline2, m_nThickness, list);

                for (int i = m_nThickness; i >= 1; --i)
                {
                    System.Drawing.Color clr = list[i - 1];
                    using (Pen pen = new Pen(clr, i))
                    {
                        pen.LineJoin = LineJoin.Round;
                        pGraphics.DrawPath(pen, pPath);
                    }
                }

                if (m_bClrText)
                {
                    using (SolidBrush brush = new SolidBrush(m_clrText))
                    {
                        pGraphics.FillPath(brush, pPath);
                    }
                }
                else
                {
                    pGraphics.FillPath(m_brushText, pPath);
                }
            }
			return true;
		}

		protected System.Drawing.Color m_clrText;
		protected System.Drawing.Color m_clrOutline1;
		protected System.Drawing.Color m_clrOutline2;
		protected System.Drawing.Brush m_brushText;
		protected bool m_bClrText;
        protected bool m_bUseCurvedGradient;
        protected bool disposed;
	}
}

File: OutlineText/TextDesigner/TextGradOutlineStrategy.cpp
/*
Text Designer Outline Text Library 

Copyright (c) 2009 Wong Shao Voon

The Code Project Open License (CPOL)
http://www.codeproject.com/info/cpol10.aspx
*/

#include "StdAfx.h"
#include "TextGradOutlineStrategy.h"
#include "TextGradOutlineLastStrategy.h"
#include "GDIPath.h"

using namespace TextDesigner;

TextGradOutlineStrategy::TextGradOutlineStrategy(void)
:
m_pbrushText(NULL),
m_bClrText(true),
m_bUseCurvedGradient(false)
{
}

TextGradOutlineStrategy::~TextGradOutlineStrategy(void)
{
}

ITextStrategy* TextGradOutlineStrategy::Clone()
{
	TextGradOutlineStrategy* p = new TextGradOutlineStrategy();
	if(m_bClrText)
		p->Init(m_clrText, m_clrOutline1, m_clrOutline2, m_nThickness, m_bUseCurvedGradient);
	else
		p->Init(m_pbrushText, m_clrOutline1, m_clrOutline2, m_nThickness, m_bUseCurvedGradient);

	return static_cast<ITextStrategy*>(p);
}

void TextGradOutlineStrategy::Init(
	Gdiplus::Color clrText, 
	Gdiplus::Color clrOutline1, 
	Gdiplus::Color clrOutline2, 
	int nThickness,
	bool useCurveGradient)
{
	m_clrText = clrText; 
	m_bClrText = true;
	if(clrOutline1.GetAlpha() < 255)
		m_clrOutline1 = Gdiplus::Color(255, clrOutline1.GetRed(), clrOutline1.GetGreen(), clrOutline1.GetBlue() );
	else
		m_clrOutline1 = clrOutline1; 

	if(clrOutline2.GetAlpha() < 255)
		m_clrOutline2 = Gdiplus::Color(255, clrOutline2.GetRed(), clrOutline2.GetGreen(), clrOutline2.GetBlue() );
	else
		m_clrOutline2 = clrOutline2; 

	m_nThickness = nThickness; 
	m_bUseCurvedGradient = useCurveGradient;
}

void TextGradOutlineStrategy::Init(
	Gdiplus::Brush* pbrushText, 
	Gdiplus::Color clrOutline1, 
	Gdiplus::Color clrOutline2, 
	int nThickness,
	bool useCurveGradient)
{
	if(m_pbrushText&&m_pbrushText!=pbrushText)
		delete m_pbrushText;

	m_pbrushText = pbrushText; 
	m_bClrText = false;
	if(clrOutline1.GetAlpha() < 255)
		m_clrOutline1 = Gdiplus::Color(255, clrOutline1.GetRed(), clrOutline1.GetGreen(), clrOutline1.GetBlue() );
	else
		m_clrOutline1 = clrOutline1; 

	if(clrOutline2.GetAlpha() < 255)
		m_clrOutline2 = Gdiplus::Color(255, clrOutline2.GetRed(), clrOutline2.GetGreen(), clrOutline2.GetBlue() );
	else
		m_clrOutline2 = clrOutline2; 

	m_nThickness = nThickness; 
	m_bUseCurvedGradient = useCurveGradient;
}

bool TextGradOutlineStrategy::DrawString(
	Gdiplus::Graphics* pGraphics,
	Gdiplus::FontFamily* pFontFamily,
	Gdiplus::FontStyle fontStyle,
	int nfontSize,
	const wchar_t*pszText, 
	Gdiplus::Point ptDraw, 
	Gdiplus::StringFormat* pStrFormat)
{
	using namespace Gdiplus;
	GraphicsPath path;
	Status status = path.AddString(pszText,wcslen(pszText),pFontFamily,fontStyle,nfontSize,ptDraw,pStrFormat);
	if(status!=Ok)
		return false;

	std::vector<Gdiplus::Color> vec;
	if (m_bUseCurvedGradient)
		TextGradOutlineLastStrategy::CalculateCurvedGradient(m_clrOutline1, m_clrOutline2, m_nThickness, vec);
	else
		TextGradOutlineLastStrategy::CalculateGradient(m_clrOutline1, m_clrOutline2, m_nThickness, vec);
	for(int i=m_nThickness; i>=1; --i)
	{
		Gdiplus::Color clr = vec.at(i-1);
		Pen pen(clr, i);
		pen.SetLineJoin(LineJoinRound);
		pGraphics->DrawPath(&pen, &path);
	}

	Status status2 = Ok;
	if(m_bClrText)
	{
		SolidBrush brush(m_clrText);
		status2 = pGraphics->FillPath(&brush, &path);
	}
	else
	{
		status2 = pGraphics->FillPath(m_pbrushText, &path);
	}

	return status2 == Ok;
}

bool TextGradOutlineStrategy::DrawString(
	Gdiplus::Graphics* pGraphics,
	Gdiplus::FontFamily* pFontFamily,
	Gdiplus::FontStyle fontStyle,
	int nfontSize,
	const wchar_t*pszText, 
	Gdiplus::Rect rtDraw, 
	Gdiplus::StringFormat* pStrFormat)
{
	using namespace Gdiplus;
	GraphicsPath path;
	Status status = path.AddString(pszText,wcslen(pszText),pFontFamily,fontStyle,nfontSize,rtDraw,pStrFormat);
	if(status!=Ok)
		return false;

	std::vector<Gdiplus::Color> vec;
	if (m_bUseCurvedGradient)
		TextGradOutlineLastStrategy::CalculateCurvedGradient(m_clrOutline1, m_clrOutline2, m_nThickness, vec);
	else
		TextGradOutlineLastStrategy::CalculateGradient(m_clrOutline1, m_clrOutline2, m_nThickness, vec);
	for(int i=m_nThickness; i>=1; --i)
	{
		Gdiplus::Color clr = vec.at(i-1);
		Pen pen(clr, i);
		pen.SetLineJoin(LineJoinRound);
		pGraphics->DrawPath(&pen, &path);
	}

	Status status2 = Ok;
	if(m_bClrText)
	{
		SolidBrush brush(m_clrText);
		status2 = pGraphics->FillPath(&brush, &path);
	}
	else
	{
		status2 = pGraphics->FillPath(m_pbrushText, &path);
	}

	return status2 == Ok;
}

bool TextGradOutlineStrategy::GdiDrawString(
	Gdiplus::Graphics* pGraphics, 
	LOGFONTW* pLogFont,
	const wchar_t*pszText, 
	Gdiplus::Point ptDraw)
{
	using namespace Gdiplus;
	Gdiplus::GraphicsPath* pPath=NULL;
	bool b = GDIPath::GetStringPath(
		pGraphics, 
		&pPath, 
		pszText, 
		pLogFont,
		ptDraw);

	if(false==b)
	{
		if(pPath)
		{
			delete pPath;
			pPath = NULL;
		}
		return false;
	}

	std::vector<Gdiplus::Color> vec;
	if (m_bUseCurvedGradient)
		TextGradOutlineLastStrategy::CalculateCurvedGradient(m_clrOutline1, m_clrOutline2, m_nThickness, vec);
	else
		TextGradOutlineLastStrategy::CalculateGradient(m_clrOutline1, m_clrOutline2, m_nThickness, vec);
	for(int i=m_nThickness; i>=1; --i)
	{
		Gdiplus::Color clr = vec.at(i-1);
		Pen pen(clr, i);
		pen.SetLineJoin(LineJoinRound);
		pGraphics->DrawPath(&pen, pPath);
	}

	Status status2 = Ok;
	if(m_bClrText)
	{
		SolidBrush brush(m_clrText);
		status2 = pGraphics->FillPath(&brush, pPath);
	}
	else
	{
		status2 = pGraphics->FillPath(m_pbrushText, pPath);
	}

	if(pPath)
	{
		delete pPath;
		pPath = NULL;
	}

	return status2 == Ok;
}

bool TextGradOutlineStrategy::GdiDrawString(
	Gdiplus::Graphics* pGraphics, 
	LOGFONTW* pLogFont,
	const wchar_t*pszText, 
	Gdiplus::Rect rtDraw)
{
	using namespace Gdiplus;
	Gdiplus::GraphicsPath* pPath=NULL;
	bool b = GDIPath::GetStringPath(
		pGraphics, 
		&pPath, 
		pszText, 
		pLogFont,
		rtDraw);

	if(false==b)
	{
		if(pPath)
		{
			delete pPath;
			pPath = NULL;
		}
		return false;
	}

	std::vector<Gdiplus::Color> vec;
	if (m_bUseCurvedGradient)
		TextGradOutlineLastStrategy::CalculateCurvedGradient(m_clrOutline1, m_clrOutline2, m_nThickness, vec);
	else
		TextGradOutlineLastStrategy::CalculateGradient(m_clrOutline1, m_clrOutline2, m_nThickness, vec);
	for(int i=m_nThickness; i>=1; --i)
	{
		Gdiplus::Color clr = vec.at(i-1);
		Pen pen(clr, i);
		pen.SetLineJoin(LineJoinRound);
		pGraphics->DrawPath(&pen, pPath);
	}

	Status status2 = Ok;
	if(m_bClrText)
	{
		SolidBrush brush(m_clrText);
		status2 = pGraphics->FillPath(&brush, pPath);
	}
	else
	{
		status2 = pGraphics->FillPath(m_pbrushText, pPath);
	}

	if(pPath)
	{
		delete pPath;
		pPath = NULL;
	}
	return status2 == Ok;
}

